components/list/list-header.js

```diff
@@ -129,28 +157,88 @@ class ListHeader extends RtlMixin(LocalizeCoreElement(LitElement)) {
 	render() {
 		const classes = {
 			'd2l-list-header-container': true,
-			'd2l-list-header-container-slim': (!this._hasActions && !this.selectAllPagesAllowed),
+			'd2l-list-header-container-pre-showing': (this.requiresSelection && this._state === 'pre-showing'),
+			'd2l-list-header-container-showing': (this.requiresSelection && this._state === 'showing'),
+			'd2l-list-header-container-hiding': (this._state === 'hiding')
+		};
+		const classesInner = {
+			'd2l-list-header-container-inner': true,
+			'd2l-list-header-slim': (!this._hasActions && !this.selectAllPagesAllowed),
 			'd2l-list-header-extend-separator': this._extendSeparator
 		};
 		return html`
 			<div class="${classMap(classes)}">
-				${this.noSelection ? null : html`
-					<d2l-selection-select-all></d2l-selection-select-all>
-					<d2l-selection-summary
-						aria-hidden="true"
-						class="d2l-list-header-summary"
-						no-selection-text="${this.localize('components.selection.select-all')}">
-					</d2l-selection-summary>
-					${this.selectAllPagesAllowed ? html`<d2l-selection-select-all-pages></d2l-selection-select-all-pages>` : null}
-				`}
-				<div class="d2l-list-header-actions">
-					<d2l-overflow-group opener-type="icon"><slot @slotchange="${this._handleSlotChange}"></slot></d2l-overflow-group>
+				<div class="${classMap(classesInner)}">
+					${this.noSelection ? null : html`
+						<d2l-selection-select-all></d2l-selection-select-all>
+						<d2l-selection-summary
+							aria-hidden="true"
+							class="d2l-list-header-summary"
+							no-selection-text="${this.localize('components.selection.select-all')}">
+						</d2l-selection-summary>
+						${this.selectAllPagesAllowed ? html`<d2l-selection-select-all-pages></d2l-selection-select-all-pages>` : null}
+					`}
+					<div class="d2l-list-header-actions">
+						<d2l-overflow-group opener-type="icon"><slot @slotchange="${this._handleSlotChange}"></slot></d2l-overflow-group>
+					</div>
 				</div>
 			</div>
 			${!this.noSticky ? html`<div class="d2l-list-header-shadow"></div>` : null}
 		`;
 	}
 
+	updated(changedProperties) {
+		super.updated(changedProperties);
+		if (!this.requiresSelection || !changedProperties.has('selectionInfo')) return;
+
+		const hasSelection = (this.selectionInfo.state !== 'none');
+
+		if (hasSelection) {
+			if (this._state === 'hiding') {
+				this._state = 'showing';
```

Comment:
Do we not need to cancel the 'transitionend' listener in this case?

Reply:
I tried to break it in testing, but I was unsuccessful. But yeah, probably to be safe, because it does seem like there is an edge case where we could see some very buggy behaviour if the state was toggled at the exact "wrong" time.